### ISSUE TITLE
[Unit Test]: Digest Algorithm update for Authenticated Variables (Bugzilla Bug 2949)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#9975]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 2949
Date: 2020-09-01T15:13:39+00:00
From: Divneil Wadhawan &lt;<divneil.r.wadhawan>&gt;
To:   Divneil Wadhawan &lt;<divneil.r.wadhawan>&gt;
CC:   @out0xb2, @jyao1

Blocker for:  #2875
See also:     https://bugzilla.tianocore.org/show_bug.cgi?id=2979
Last updated: 2021-04-26T20:48:39+00:00

**Comment by tianocore-issues** [comment]
## Comment 13596
Date: 2020-09-01 15:13:39 +0000
From: Divneil Wadhawan &lt;<divneil.r.wadhawan>&gt;

- Industry Specification: UEFI Specification
- Target OS: ---
- Bugzilla Assignee(s): Divneil Wadhawan &lt;<divneil.r.wadhawan>&gt;

This is to track unit test development for Authenticated Variable with EmulatorPkg

**Comment by tianocore-issues** [comment]
## Comment 13602
Date: 2020-09-02 08:28:34 +0000
From: Divneil Wadhawan &lt;<divneil.r.wadhawan>&gt;

The following 2 patches enable Authenticated Variable Unit Test in EmulatorPkg

a. Enable AuthVar support in Emulator Pkg
   https://github.com/divneil/edk2-staging/pull/new/Bugzilla2949-enable-Authenticated-Variable-Support-in-EmulatorPkg

b. Unit Test for AuthVar
   https://github.com/divneil/edk2-staging/pull/new/Bugzilla2949-Unit-Test-for-Authenticated-Variables

   This contains the unit-test case of &apos;PK&apos; update with the following algorithms:
   a. RSA2048/SHA256
   b. RSA3072/SHA256
   c. RSA4096/SHA256
   d. RSA3072/SHA384 # AuthVar Update support is in Bug2950
   e. RSA4096/SHA512 # AuthVar Update support is in Bug2950

**Comment by tianocore-issues** [comment]
## Comment 13720
Date: 2020-09-21 03:40:14 +0000
From: Divneil Wadhawan &lt;<divneil.r.wadhawan>&gt;

(In reply to Divneil Wadhawan from comment #1)
&gt; The following 2 patches enable Authenticated Variable Unit Test in
&gt; EmulatorPkg
&gt; 
&gt; a. Enable AuthVar support in Emulator Pkg
&gt;   
&gt; https://github.com/divneil/edk2-staging/pull/new/Bugzilla2949-enable-
&gt; Authenticated-Variable-Support-in-EmulatorPkg

This patch has been modified to enable SECURE BOOT in EmulatorPkg. Following are the details as of now.

a. Enable Secure boot in EmulatorPkg
   https://edk2.groups.io/g/devel/message/65335 (Merged)

b. Due to cross between patches as detailed in https://edk2.groups.io/g/devel/message/65405, a new patch has been generated in https://edk2.groups.io/g/devel/message/65410 and is under review.

c. Bug2979 has been raised for enabling SECURE BOOT CI in EmulatorPkg.

**Comment by tianocore-issues** [comment]
## Comment 13721
Date: 2020-09-21 13:22:59 +0000
From: Divneil Wadhawan &lt;<divneil.r.wadhawan>&gt;

(In reply to Divneil Wadhawan from comment #2)

&gt; This patch has been modified to enable SECURE BOOT in EmulatorPkg. Following
&gt; are the details as of now.
&gt; 
&gt; a. Enable Secure boot in EmulatorPkg
&gt;    https://edk2.groups.io/g/devel/message/65335 (Merged)
&gt; 
&gt; b. Due to cross between patches as detailed in
&gt; https://edk2.groups.io/g/devel/message/65405, a new patch has been generated
&gt; in https://edk2.groups.io/g/devel/message/65410 and is under review.
Merged on edk2/master

&gt; b. Unit Test for AuthVar
&gt;   
&gt; https://github.com/divneil/edk2-staging/pull/new/Bugzilla2949-Unit-Test-for-
&gt; Authenticated-Variables
&gt; 
&gt;    This contains the unit-test case of &apos;PK&apos; update with the following
&gt; algorithms:
&gt;    a. RSA2048/SHA256
&gt;    b. RSA3072/SHA256
&gt;    c. RSA4096/SHA256
&gt;    d. RSA3072/SHA384 # AuthVar Update support is in Bug2950
&gt;    e. RSA4096/SHA512 # AuthVar Update support is in Bug2950
The directory of the unit test has been changed from UnitTestFrameworkPkg to MdeModulePkg. A new branch is created:
https://github.com/divneil/edk2-staging/pull/new/Bugzilla2949-Unit-Test-for-Authenticated-Variables-MdeModulePkg

**Comment by tianocore-issues** [comment]
## Comment 15762
Date: 2021-04-22 09:21:24 +0000
From: @out0xb2

Let me start by saying, I really like the progress here!

I observe that this test is testing 2 properties unrelated to the digest on the authenticated variable signature, and that may confuse people?

(1) Certificate key size
(2) Certificate digest algorithm

The key size on the certificates is unrelated to the proposed code and spec change.  The digest algorithm used in the certificates are unrelated to the proposed code and spec change.  For example, if a certificate is generated using RSA-4096 and SHA-512 digests, and then this certificate signed an authenticated variable using SHA2-256 digest (this is permitted), that would work today without the code change to TianoCore (I tested it a couple weeks ago).  This is because EDK2 does not look at the certificate key size or certificate digest algorithm, that is simply handled by OpenSSL.  EDK2 only looks at the digest alg on the outer signature.

So, what does Jeremiah suggest?

The digest algorithm on the signature should be a separate parameter from the digest algorithm used when creating certificates.  A test for signature digest algorithms should select a certificate, then iterate over signature digest algorithms.  If we want to test certificate key sizes and certificate digest algorithms, which is really a test of our usage of OpenSSL (and that is good), but I think that should be a separate test.

In a nutshell, split into a signature digest test, and a certificate test that includes various key sizes and digest algorithms.

**Comment by tianocore-issues** [comment]
## Comment 15763
Date: 2021-04-22 11:13:35 +0000
From: @jyao1

Agree with Jeremiah.

Digest algorithm in certificate is totally different with digest algorithm in P7 signed data.

The code https://github.com/tianocore/edk2/blob/master/SecurityPkg/Library/AuthVariableLib/AuthService.c#L1917 handles mSha256OidValue only.

We need extend them to handle more such as mSha384OidValue, mSha512OidValue.

If the purpose of this Bugzilla is to test different digest size of P7 data, then we need do more.

**Comment by tianocore-issues** [comment]
## Comment 15769
Date: 2021-04-25 09:00:25 +0000
From: Divneil Wadhawan &lt;<divneil.r.wadhawan>&gt;

(In reply to Jeremiah Cox from comment #4)
&gt; I observe that this test is testing 2 properties unrelated to the digest on
&gt; the authenticated variable signature, and that may confuse people?
&gt; 
&gt; (1) Certificate key size
&gt; (2) Certificate digest algorithm

---
./gen_auth_var.sh --rsa=2048 --digest=256
./gen_auth_var.sh --rsa=3072 --digest=256
./gen_auth_var.sh --rsa=4096 --digest=256

./gen_auth_var.sh --rsa=3072 --digest=384
./gen_auth_var.sh --rsa=4096 --digest=512
---

o The first 3 test cases were written to make sure that no regression has happened due to this change. I have come across many people referring this https://wiki.archlinux.org/index.php/Unified_Extensible_Firmware_Interface/Secure_Boot to generate the secure boot key hierarchy, and there&apos;s a 1:1 mapping of CertDigestAlgo and SignedDataDigestAlgo so, I thought this will bring in parity.

o The next 2 test cases, followed the same approach of 1:1 mapping. However, it does test the multiple SignedDataDigestAlgo, but I understand it is presenting a 1:1 relationship between CertDigestAlgo and SignedDataDigestAlgo, which isn&apos;t correct.

So, as I understand you are looking for:
for every CertDigestAlgo, SignedDataDigestAlgo = {256, 384, 512) to be tested.

(In reply to jiewen.yao from comment #5)
&gt;
&gt; The code
&gt; https://github.com/tianocore/edk2/blob/master/SecurityPkg/Library/
&gt; AuthVariableLib/AuthService.c#L1917 handles mSha256OidValue only.

The changes are already in https://github.com/divneil/edk2-staging/pull/new/Bugzilla2950-AuthService-extending-digest-algorithm-support (Bug2950)

**Comment by tianocore-issues** [comment]
## Comment 15780
Date: 2021-04-26 20:48:39 +0000
From: @out0xb2

To me, the important part is to separate cert digest from signature digest.  One approach that keeps things untangled is to have 3 separate tests or test engines, each of which focuses on testing 1 degree of freedom while keeping the others the same:  key size, cert digest, signature digest.  But I do see the attraction of having 1 test engine that perhaps has a nested for loop, where each nest covers an  aspects (key size, cert digest, signature digest).  If this nested for loop was named and documented well, that would probably satisfy my concern and be most comprehensive.

Thank you for considering this feedback, we do appreciate the increase in tests and test coverage (and crypto algorithms).
